### ISSUE TITLE
Retain renderCell state after paginating data in DataGrid

### ISSUE THREAD

**Issue by ShadiBlitz** [mui/mui-x#4815]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Moving pages resets the renderCell component to their initial state.

### Steps to reproduce 🕹

For example make a switch componenet and add it to the render cell, make some switches on and move to next page.
The switches in the previous page will be back to their initial states.

https://codesandbox.io/s/datagriddemo-material-demo-forked-ygrpw6?file=/demo.tsx

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by alexfauquette** [comment]
That's the expected behavior. When a component is not visible it is not rendered to improve performances.

Other cells are consistent between pages because we internally save data modification, and you should be doing the same. 

Maybe saving the state of the switches in rows, or in another state could be nice?. For example see the `toggleAdmin` in the `'action'` column example
https://mui.com/x/react-data-grid/columns/#special-properties

**Comment by ShadiBlitz** [comment]
is there a better solution for a renderCell component? because all the login and state relies inside it, if I try to move the state outside it won't work.

**Comment by ShadiBlitz** [comment]
> That's the expected behavior. When a component is not visible it is not rendered to improve performances.
> 
> Other cells are consistent between pages because we internally save data modification, and you should be doing the same.
> 
> Maybe saving the state of the switches in rows, or in another state could be nice?. For example see the `toggleAdmin` in the `'action'` column example https://mui.com/x/react-data-grid/columns/#special-properties

is there a better solution for a renderCell component? because all the login and state relies inside it, if I try to move the state outside it won't work.

**Comment by alexfauquette** [comment]
I don't get why it won't work. Could provide more context about what you are trying to build?

**Comment by ShadiBlitz** [comment]
My columns and rows are separated each in their own files and exported.
My DataGrid component imports them in the main page, and in the column, there is a renderCell that includes a component that has it's own states and logic. How do I manage to not rerender the DataGrid so I can have the renderCell component state saved? The above solution is not practical because rows and columns are separated and each have their own logic.

**Comment by alexfauquette** [comment]
The easiest way would be to save data in the internal state of the grid as follow

https://codesandbox.io/s/datagriddemo-material-demo-forked-pgxi0z?file=/demo.tsx

**Comment by flaviendelangle** [comment]
But updating rows like this triggers the sorting/ filtering process so unless you have a very limited amount of rows it will be very heavy.

**Comment by ShadiBlitz** [comment]
> 

Agreed, do you have a solution of how such thing could be approached?

**Comment by ShadiBlitz** [comment]
> But updating rows like this triggers the sorting/ filtering process so unless you have a very limited amount of rows it will be very heavy.

Agreed, do you have a solution of how such thing could be approached?

**Comment by flaviendelangle** [comment]
You could have a ref to store those information [like on this example](https://codesandbox.io/s/datagriddemo-material-demo-forked-g808ri?file=/demo.tsx)

To be honest it's not a perfect solution, storing render information in a ref is not great.
We could dive deeper to see if we could provide a built-in api.

**Comment by flaviendelangle** [comment]
Closes by #5158

**Comment by scamden** [comment]
from the PR and docs (which i'm scouring)
> you should save it either in the data grid state

i don't think it's clear how to do this. i seein the codepen above the "state" being stored is on the row, but in my case i rely on some state that isn't rowwise. it's more like general metadata. is there an api to store generalized state, or state on a column? @flaviendelangle

**Comment by flaviendelangle** [comment]
We are both not working on the data grid anymore.
I'll let @MBilalShafi help you on that one :+1:

**Comment by MBilalShafi** [comment]
Hey @scamden

> it's more like general metadata

Could you explain what kind of metadata it is and how you are using it with `renderCell`? A codesandbox example would be perfect. Could you use `ref` based solution as suggested in https://github.com/mui/mui-x/issues/4815#issuecomment-1127620929 ?

**Comment by scamden** [comment]
I'll provide an example but abstractly i guess i just wondered what the meaning of "saving in grid state" is since that docs mention it. Does this simply mean on a row?

The type of metadata i mean is something like a column definition. Hypothetically: a `format` string that the renderCell will use to do it's work. These are passed into the Grid's parent component as props and captured in the closure of renderCell. Do I need to calling apRef.current.updateColumn in an effect for these? (Or what's the general pattern for when column definitions need to be updated reactively)

**Comment by scamden** [comment]
i should mentioned we are a premium customer

**Comment by scamden** [comment]
https://codesandbox.io/s/heuristic-butterfly-nc3khc?file=/src/Demo.tsx

here's an example of what i mean, i filed an official [issue](https://github.com/mui/mui-x/issues/10895) for this cause i believe it's a bug actually.

**Comment by scamden** [comment]
in general it would be nice to attach state to a particular column which could get updated separately from the colDef
